lib/web/apiserver.go

```diff
@@ -4091,11 +4091,38 @@ func (h *Handler) AuthenticateRequest(w http.ResponseWriter, r *http.Request, ch
 		if err != nil {
 			return nil, trace.AccessDenied("need auth")
 		}
-		if err := ctx.validateBearerToken(r.Context(), creds.Password); err != nil {
+		if err := sctx.validateBearerToken(r.Context(), creds.Password); err != nil {
 			return nil, trace.AccessDenied("bad bearer token")
 		}
 	}
-	return ctx, nil
+
+	if err := parseMFAResponseFromRequest(r); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return sctx, nil
+}
+
+// parseMFAResponseFromRequest attempts to parse an MFA reponse from the request body.
+// If found, the MFA reponse is added to the request context where it can be recalled
+// further down the call stack.
+func parseMFAResponseFromRequest(r *http.Request) error {
+	if mfaResponseJSON := r.Header.Get("Mfa-Response"); mfaResponseJSON != "" {
+		var resp mfaResponse
+		if err := json.Unmarshal([]byte(mfaResponseJSON), &resp); err != nil {
+			return trace.Wrap(err)
+		}
+
+		mfaResp := &proto.MFAAuthenticateResponse{
+			Response: &proto.MFAAuthenticateResponse_Webauthn{
+				Webauthn: wantypes.CredentialAssertionResponseToProto(resp.WebauthnAssertionResponse),
+			},
+		}
+
+		ctx := mfa.ContextWithMFAResponse(r.Context(), mfaResp)
+		*r = *r.WithContext(ctx)
```

Comment:
They are necessary to mutate the request with the updated context where it can be used higher up the callstack.

Another option would be to move this logic out of the Auth handler and instead handle it within individual handlers:

```go
func (h *Handler) updateUserHandle(w http.ResponseWriter, r *http.Request, params httprouter.Params, ctx *SessionContext) (interface{}, error) {
	clt, err := ctx.GetClient()
	if err != nil {
		return nil, trace.Wrap(err)
	}

	ctx := contextWithMFAResponseFromHeader(r.Context(), r.Header)
	return updateUser(ctx, r, clt, ctx.GetUser())
}
```